calico/openstack/mech_calico.py

```diff
@@ -100,26 +103,52 @@ def __init__(self):
         # initialize these properly when we first need them.
         self.db = None
         self.transport = None
-        self._state_initialized = False
+        self._my_pid = None
+        self._periodic_resync_greenlet = None
+        self._epoch = 0
 
-        # Start our resynchronization process.
-        eventlet.spawn(self.periodic_resync_thread)
+        # Make sure we initialise even if we don't see any API calls.
+        eventlet.spawn_after(STARTUP_DELAY_SECS, self._init_state)
 
     def _init_state(self):
         """
         Creates the connection state required for talking to the Neutron DB
         and to etcd. This is a no-op if it has been executed before.
         """
-        if self._state_initialized:
+        current_pid = os.getpid()
+        if self._my_pid == current_pid:
+            # We've initialised our PID and it hasn't changed since last time,
+            # nothing to do.
```

Comment:
It is a bit paranoid, the main change here is to initialise the PR thread from the lazy init (which was already being called on every API call) rather than the other way around.  We still don't get a PR thread in every child until it receives an API call but the parent always starts one.  I can't see a way around that unless we can get a post-fork kick.

I think the cost of field access vs getpid is negligible.  If we want to remove the overhead of the lazy-init decorator we'd have to have it patch itself out after it's been called.  I.e. replace the entry in `__dict__` with the rawmethod without decorator.   I expect that's not worth it.

Hmm, maybe there's a generic python/syscall API for registering a post-fork callback?  Then we wouldn't have to get a kick plumbed through Neutron.  Or, we could set up our own IPC channel of some sort before the fork so that parent can prod children.